service/matching/matcher_test.go

```diff
@@ -211,54 +211,57 @@ func (t *MatcherTestSuite) testRemoteSyncMatch(taskSource enumsspb.TaskSource) {
 }
 
 func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
-	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	intruptC := make(chan struct{})
 
-	// sync match happens when there is no backlog
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	go func() { historyTask.responseC <- nil }()
-	happened, err := t.childMatcher.Offer(ctx, historyTask)
-	t.True(happened)
-	t.Nil(err)
+	// task waits for a local poller
+	oldBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
 
-	intruptC := make(chan struct{})
-	youngBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Second), nil)
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
-	time.Sleep(time.Millisecond)
+	var wg sync.WaitGroup
+	go func() {
+		t.rootMatcher.MustOffer(ctx, oldBacklogTask, intruptC) //nolint:errcheck
+	}()
 
-	// should allow sync match when there is only young tasks in the backlog
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
-	t.Nil(err)
-	t.False(happened) // sync match did not happen, but we called the forwarder client
+	// Wait for the task to be added to the map
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			if len(t.rootMatcher.backlogTasksCreateTime) > 0 {
+				return // Exit if a task has been added
+			}
+		}
+	}()
+	wg.Wait()
```

Comment:
We need this check to be done - I ran this without a wait group which verifies if the map has increased in length or not and the test began to flake when ran 60,000 times. This cleared multiple times when run the same number of times.